### ISSUE TITLE
Current make-hook/make-handler practice is too rigid?

### ISSUE THREAD

**Issue by BlueFlo0d** [atlas-engineer/nyxt#1966]
It seems that currently, all those functions accept function objects instead of symbols.
Therefore there's no easy way to redefine a handler function, which makes it very difficult to develop on a long running image. Am I missing anything?

**Comment by aartaka** [comment]
> It seems that currently, all those functions accept function objects instead of symbols.
> Therefore there's no easy way to redefine a handler function, which makes it very difficult to develop on a long running image. Am I missing anything?

There is a way -- simply do `hooks:remove-hook` and pass the name of the handler to it (that's the whole purpose for the named handlers). After it's removed, you can `hooks:add-hook` again, thus updating the named hook.

**Comment by BlueFlo0d** [comment]
> 
> There is a way -- simply do `hooks:remove-hook` and pass the name of the handler to it (that's the whole purpose for the named handlers). After it's removed, you can `hooks:add-hook` again, thus updating the named hook.

However hooks are not named (and we can't name them because almost all hooks are buffer local). One need to iterate over all buffer to update a hook this way. IMO it's too much a hassle comparing to just add one needed level of indirection. What do you think?

In general, I'm not sure if `serapeum:hooks` is the right thing for our use cases. We have very few use cases for anonymous function handlers (I counted 4), and all of them just need a little bit of rewrite to turn into named function handlers by turning closure variables into buffer-local-variable, and by doing so we make them much more transparent. In the end we just need something as simple as
```
(defclass hook () ((functions) (combination))
```
Just one `combination` more than what we do in Emacs.

**Comment by aartaka** [comment]
> However hooks are not named (and we can't name them because almost all hooks are buffer local). One need to iterate over all buffer to update a hook this way. IMO it's too much a hassle comparing to just add one needed level of indirection. What do you think?

Sorry, hooks are not named indeed, I've been unclear about it. I meant named handlers — those are easily removable. While your point on buffer iteration is valid, destructively modifying buffer-local hooks is not an intended use-case. They are meant to be `define-configuration`-ed, thus allowing both universality (all the instances of a class get a reasonably similar value) and ability to revert to the older value (due to our user-classes system).
 
> In general, I'm not sure if `serapeum:hooks` is the right thing for our use cases.

It's fun that you put it this way, because `serapeum/contrib/hooks` originated and was made exactly for Nyxt :D

> We have very few use cases for anonymous function handlers (I counted 4), and all of them just need a little bit of rewrite to turn into named function handlers by turning closure variables into buffer-local-variable, and by doing so we make them much more transparent.

While Nyxt source code lacks unnamed handlers, it's not always the same about user configuration code. Writing a simple lambda and giving it a name is simpler than doing a `defun`, and is sometimes the right choice. Especially for the throwaway actions that wouldn't mean anything as toplevel functions.

Introducing additional slots for every hook would lead to an unnecessary growth of code. Using features of the language, such as closures, is perfectly fine for such use-cases.

> In the end we just need something as simple as
> 
> ```
> (defclass hook () ((functions) (combination))
> ```
> 
> Just one `combination` more than what we do in Emacs.

Emacs is not to be a role-model here. The hooks system of Nyxt was designed to avoid the problems that Emacs hooks had. And Nyxt hooks system does work and is flexible enough to allow most of possible use-cases, especially when paired with our configuration system.

**Comment by BlueFlo0d** [comment]
> While your point on buffer iteration is valid, destructively modifying buffer-local hooks is not an intended use-case. They are meant to be define-configuration-ed, thus allowing both universality (all the instances of a class get a reasonably similar value) and ability to revert to the older value (due to our user-classes system).

The same problem happens to define-configuration-ed hooks. Redefining the configuration to a different value doesn't affect existing buffers, and one have to either kill and reopen all buffers, or iterate through them (both needs de-facto iteration).

I also don't think such use-case should be designed as "unintended", if one want a flexible and long-running operating environment. Nyxt is sold as "hacker-oriented".

> While Nyxt source code lacks unnamed handlers, it's not always the same about user configuration code. Writing a simple lambda and giving it a name is simpler than doing a defun, and is sometimes the right choice. Especially for the throwaway actions that wouldn't mean anything as toplevel functions.
First, I doubt the claim that "writing a simple lambda and giving it a name is simpler than doing a defun". The former is `(lambda (...) ...) :name name`, the latter is `(defun name (...) ...)`. They are arguably equally simple, but the latter has the benefit of easy redefinition and access (e.g. play and test in REPL). Also the user don't necessarily need to give a function any meaning when defining it -- it's very typical that someone wrote some throwaway Emacs Lisp functions, and they eventually evolve into a whole package. This is possible because most Emacs functions are easily accessed and redefined (thus evolvable). 

> Introducing additional slots for every hook would lead to an unnecessary growth of code. Using features of the language, such as closures, is perfectly fine for such use-cases.
I count only 4 (2 of them are really similar thus de-facto only 3) uses of closure handlers in Nyxt code base. There's a tradeoff between closure/buffer slot, but I think the introduction of code growth is minimal and worth the additional accessibility and transparency it gives. In CL it's in general hard to inspect and hack a closure.

All I'm saying does not mean we shall ban anonymous function. The simplest implementation that accepts symbols/named functions will also accept unnamed functions. But named functions should be encouraged. This also implies that the documentation/type slots are probably unnecessary. For named functions, they can be retrieved from the function symbol, and for unnamed "throw away" functions in your conception, they don't need them anyway.

> Emacs is not to be a role-model here. The hooks system of Nyxt was designed to avoid the problems that Emacs hooks had. And Nyxt hooks system does work and is flexible enough to allow most of possible use-cases, especially when paired with our configuration system.
From my arguments above, it also introduces additional problem, especially significant loss of accessibility and transparency, which we should not be shame about learning back from Emacs.

I think I'm ready to summarize my proposals at this point. There're several points, and I'd love each to be assessed individually. They definitely don't need to be accepted or rejected as a whole. 
1. `hooks` facility will accept symbols in addition to function objects.
2. Nyxt code base will encourage using symbols rather than anonymous function objects. The ubiquitous use of `:handlers (list #'...)` will be updated to `:handlers (list '...)`
3. `handler` class will be simplified. `name`, `description`, `handler-type` will be removed because they're either unnecessary for throw away function, or can already be retrieved from function symbol.
4. The few use cases of closure handlers in Nyxt will be replaced by named function + buffer slots.

Does some of the above make sense?

**Comment by aartaka** [comment]
> I think I'm ready to summarize my proposals at this point. There're several points, and I'd love each to be assessed individually. They definitely don't need to be accepted or rejected as a whole.
> 
> 1. `hooks` facility will accept symbols in addition to function objects.
> 2. Nyxt code base will encourage using symbols rather than anonymous function objects. The ubiquitous use of `:handlers (list #'...)` will be updated to `:handlers (list '...)`
> 3. `handler` class will be simplified. `name`, `description`, `handler-type` will be removed because they're either unnecessary for throw away function, or can already be retrieved from function symbol.
>    3'. Alternatively, we can support using symbols/functions directly as handler instead of a separate `handler` class. Most use cases will use symbols directly. The `handler` class may still be retained for use with "named" anonymous function (which I don't really see any use case) or setf functions (I don't see any use case in Nyxt yet, but there could be).
> 4. The few use cases of closure handlers in Nyxt will be replaced by named function + buffer slots.
> 
> Does some of the above make sense?

Points 1 and 3' do make sense, but others I don't fully understand/agree with.
2. Why encourage symbols, actually? Is there any benefit? Function objects are much better as things to build handlers out of, as they have both type information and actually represent the computation to happen when the hook fires.
3.1. Why remove `name`? It's useful for debugging purposes.
3.2. Why remove `description`? I mean, it's totally optional, but if one can document things in a way that system allows it, why use something else?
3.3. `handler-type` can be derived from function symbol, but functions sometimes have unpredictable types like `(function ((or null integer string)) (values (or null string) &optional)` which is not exactly nice for configuration and does not necessarily work for all the functions we might bind to the hook. Moreover, some implementations are relatively limited as ftype guessing, and we won't be able to get any useful information out of those. Better use explicit types.

4 I responded to already.

I mean, your vision of using symbols due to their mutability is understandable. I'm just not sure how well it plays with the rest of Nyxt. And I'm afraid I'm currently a bit short-sighted (for a range of reasons) to discuss all the implications of the possible change. I suggest to wait until @Ambrevar and @jmercouris can join the discussion.

**Comment by BlueFlo0d** [comment]
> 2. Why encourage symbols, actually? Is there any benefit? Function objects are much better as things to build handlers out of, as they have both type information and actually represent the computation to happen when the hook fires. 

Lisp function themselves don’t have type information. If you are talking about annotations, it’s even easier to attach them to symbols using symbol properties. All your claimed benefit of functions applies equally and maybe even better for function symbols. Plus, function symbols have the vital benefit of being globally addressable and redefinable, which every Lisp hacker understands. Why does the default way to call Lisp functions `(function . args)` has the semantics of `(funcall 'function . args)`, not `(funcall #'function . args)`?

> 3.1. Why remove `name`? It's useful for debugging purposes. 3.2. Why remove `description`? I mean, it's totally optional, but if one can document things in a way that system allows it, why use something else? 

They are unnecessary for named functions because you can already retrieve them from the symbol. Now I do think 3' is better because there's little need for a separate handler class if one can just use symbol. The handler class could be retained for backward compatibility.

> 3.3. `handler-type` can be derived from function symbol, but functions sometimes have unpredictable types like `(function ((or null integer string)) (values (or null string) &optional)` which is not exactly nice for configuration and does not necessarily work for all the functions we might bind to the hook. Moreover, some implementations are relatively limited as ftype guessing, and we won't be able to get any useful information out of those. Better use explicit types.

This is an issue orthogonal to handlers. I think you're describing a pitfall of how existing CLs handle function types. The right thing to do is to solve it for all functions, e.g. using a symbol property to function symbols, or maintain a database for function objects. It’s weird to “solve” it only for a specific use case of functions in general.

> 4 I responded to already.

Sure, it’s a trade off. It’s a clear win in my view because I assign very high value to hackability.

> I mean, your vision of using symbols due to their mutability is understandable. I'm just not sure how well it plays with the rest of Nyxt. And I'm afraid I'm currently a bit short-sighted (for a range of reasons) to discuss all the implications of the possible change. I suggest to wait until @Ambrevar and @jmercouris can join the discussion.

Lots of your argument make perfect sense under the context of a modern corporation software engineering project, but seem to contradict basic consensus of (traditional) Lisp hacker community. Not to be offensive, but have you hacked in somewhat traditional Lisp environment, like Emacs, MIT Scheme, Lisp Machine, or even some flavor of Smalltalk? I’ve tried to explain why using symbols is important but it might not work because it’s hard to explain a common sense (within specific community). Still, let me try to summarize here:
1. It makes function redefinition consistently affects all its uses. This is essential for maintaining and updating a long running image, and image-based development. All parts of the system should be made redefinable. Positive examples: default semantics of calling a CL function, and update-instance-for-class.
1'.  Late binding is one of the core characteristic of Lisp.
2. Emacs community proves an additional benefit from 1: by encourage using named symbols rather than anonymous functions (unless they’re absolutely for one-time use only, like arguments to mapcar), casually written helper function could evolve into a full fledged package.
3. Symbol is addressable by a global name. You kind of reinvented one with handler’s name slot, but why not use symbol directly?

> I'm just not sure how well it plays with the rest of Nyxt.

If they indeed didn’t play, guess that means my hope of making Nyxt a universal Lisp environment is just a pipe dream, and I would stop. The part of codebase I’ve read does not deviate so much from the late-binding tradition of Lisp actually (hook being the first one which hits me), but maybe your team intend to change it, I don’t know.

**Comment by jmercouris** [comment]
I appreciate the thoughtful discussion and the proposal. There are many considerations and we will likely adapt many of the concepts you've expressed. The greater problem you hint at (of general flexibility of a running system) is one that we've sought to address.

The challenge with this problem has defined almost every one of our important architectural decisions. Why do we have objects and slots rather then def parameters? Why do we have a define-configuration macro? We are trying to avoid global state. Introducing symbols and function objects via the defun macro is doing precisely that. Introducing global state. From your perspective this side effect is incredibly desirable for hackability. This is true. There are however other considerations: how stable is our system? How fragile are user configurations? Extensions that add hook handlers, will they interfere with each other?

We are weighing many aspects of a good system to come to design decisions. With regards to my personal opinion: it should have always been possible to add a hook handler by symbol reference. If the user wishes to chain invalid combinations of handlers, that is their prerogative, to break their system in incredibly difficult to debug ways.

My final comment: I found the following paragraph unnecessary: "Lots of your argument make perfect sense under the context of a modern corporation software engineering project, but seem to contradict basic consensus of (traditional) Lisp hacker ". I interpreted it as a mode to disparage and undermine the credibility of an argument via ad hominem. It doesn't further the discussion in any meaningful way.

I am not saying Artyom is infallible, but I think he definitely meets the definition of a lisp hacker in every sense.

**Comment by BlueFlo0d** [comment]
> We are trying to avoid global state. Introducing symbols and function objects via the defun macro is doing precisely that. Introducing global state. From your perspective this side effect is incredibly desirable for hackability. This is true. There are however other considerations: how stable is our system? How fragile are user configurations? Extensions that add hook handlers, will they interfere with each other?

There are benevolent global states, like the `eq` object identity, which hardly anyone complains about. The “global state” introduced by symbol indirection is a very specific one, increasing the power of a human user while being almost transparent to most parts of the program. If a program does not manipulate obarray directly, it won’t be affect by this indirection by anyway. In fact, Smalltalk, probably the first proponent of localized state, has a similar indirection (object id table) for every objects! This makes it possible for the user to live redefine any object, but doesn’t obscure the behavior of most program (that don’t manipulate id directly) at all as their capability are still restricted by the reference graph and satisfies the strictest state localization principle.

> We are weighing many aspects of a good system to come to design decisions. With regards to my personal opinion: it should have always been possible to add a hook handler by symbol reference. If the user wishes to chain invalid combinations of handlers, that is their prerogative, to break their system in incredibly difficult to debug ways.

So at least proposal 1 is agreed on.

> I appreciate the thoughtful discussion and the proposal. There are many considerations and we will likely adapt many of the concepts you've expressed. The greater problem you hint at (of general flexibility of a running system) is one that we've sought to address.

I’m likely to do it myself if I decided continue to hack Nyxt, because it’s a blocker for my structural editor project. Can I have a list of points, which if I had implemented and opened PR, would not be rejected? It would be ideal if I can have 1 2 3’ 4, but if only part of them can be accepted by upstream, I would assess if a good universal Lisp environment in the long run is possible on the limited set and by chance could still continue the work.

> 
> My final comment: I found the following paragraph unnecessary: "Lots of your argument make perfect sense under the context of a modern corporation software engineering project, but seem to contradict basic consensus of (traditional) Lisp hacker ". I interpreted it as a mode to disparage and undermine the credibility of an argument via ad hominem. It doesn't further the discussion in any meaningful way.
> 
> I am not saying Artyom is infallible, but I think he definitely meets the definition of a lisp hacker in every sense.

If I sounds like ad hominem, I apologize. I quantified lisp hacker with “traditional” and I acknowledge lisp hackers outside those specific circles. I mean to mean what is literally stated: any meaningful discussion needs a common ground, and I found one such common ground that I presumed (the significance of symbol indirection), might does not exist. Therefore, I just want to confirm whether majority your team agree on this issue, or not. It’s not likely I can convince everyone of you by words on this aspect.

**Comment by jmercouris** [comment]
> There are benevolent global states...

I find myself in agreement. Perhaps we have been overzealous! There may be less risk in this domain than originally anticipated.

> I’m likely to do it myself if I decided continue to hack Nyxt...

I am personally in agreement with 1, 2, 3', and 4. I don't see any contradiction with our current set-up. We have the same goals, we can work on these things together and come up with an even better solution than we could alone! For example 3' was raised as an alternative to an objection, and I find it superior to 3.

> If I sounds like ad hominem, I apologize. I quantified lisp hacker with “traditional” and I acknowledge lisp hackers outside those specific circles. I mean to mean what is literally stated: any meaningful discussion needs a common ground, and I found one such common ground that I presumed (the significance of symbol indirection and late binding), might does not exist. Therefore, I just want to confirm whether majority your team agree on this issue, or not. It’s not likely I can convince everyone of you by words on this aspect.

Apology accepted, at least on my part :-D. If anyone would have taken offense it might have been Artyom (I cannot speak for him of course).

With regards to convincing us of things, I think we are quite flexible and open to discussion! We have changed many things when good arguments have been made.

P.S. why named anonymous functions? Something like the following does not make sense:

```
;; A function to update my handler which is bound to the symbol qrs
(defun xyz ()
   (defun qrs ()
      ...))
```

You cannot easily update a handler within a function body, you'll find yourself fooling around with fdefinition etc, not certain you've done everything the defun macro does. But if it is an anonymous (YET NAMED) handler, you can do it with:

```
;; A function to update my handler which is bound to an anonymous function
(defun qwe ()
   (hooks:add-hook some-hook (make-handler-* (lambda () ;; some new functionality ) :name 'name-we-can-use-for-comparison-and-updating-instances)))
```

**Comment by jmercouris** [comment]
Lastly, just so we are on the same page, so that some of our design decisions don't seem like complete nonsense, I assume you've read the following: https://nyxt.atlas.engineer/article/hooks-implementation.org ?

**Comment by BlueFlo0d** [comment]
> I am personally in agreement with 1, 2, 3', and 4. I don't see any contradiction with our current set-up. We have the same goals, we can work on these things together and come up with an even better solution than we could alone! For example 3' was raised as an alternative to an objection, and I find it superior to 3.

Great! I’ll wait for more comments, and I could implement those once I finish my vacation :D

> P.S. why named anonymous functions? Something like the following does not make sense:
> 
> ```
> ;; A function to update my handler which is bound to the symbol qrs
> (defun xyz ()
>    (defun qrs ()
>       ...))
> ```
> 

I agree this is a potential use case. 

I think updating a global handler like this is perfectly fine, though. The reason you don’t see it that often is that for a globally defined function, one hardly need to *programmatically* alternate its behavior by replacing its code. It’s usually better off just using some global variables (you might say eww global variables! It’s actually not that bad with package system and dynamic bindings, comparing to replacing function bodies). So, `qrs` will make use of some global variables, and `xyz` will toggle them on and off, etc. It’s also more discoverable and inspectable than replacing a function body. Leave that power to we humans!

There are indeed problematic cases like this for named functions, when you want the same name to denote different functionalities under different context (e.g. for different hooks). One hypothetical use case would be a general `log-me` handler which prints some specific debug information about the current hook. The root cause of this trouble is the lack of first-class environment. Maybe let me name it Greenspun’s hundredth rule:

Any sufficiently complete operating environment contains an ad hoc, informally-specified, bug-ridden, slow implementation of half of first-class environment.

As the above suggests, lots of ad hoc solutions are invented. I think we could get away with it in most cases using buffer local variables (and Nyxt’s blend of buffer slots and mode slots, which I consider a big step forward from Emacs’s buffer local variables!). Namely, similar to my example of defunctionalizing `xyz`, one can use a combination of named function and buffer/mode slots. But there could be cases that cannot be handled as such, so it’s probably a good idea to retain the handler class, and we can see if such cases actually happen in practice.

> Lastly, just so we are on the same page, so that some of our design decisions don't seem like complete nonsense, I assume you've read the following: https://nyxt.atlas.engineer/article/hooks-implementation.org ?

I’ve read the code before. I’ve just read the article, thanks for the pointer! I in fact love most part of both the code and the article.  If I’m making annoying noises about some part of a project, assume I love other parts of it — otherwise I won’t make any noise at all!

I agree with all the design choice. Only one minor detail, maybe let me number it as a proposal:

5. Disabling and then re-enabling handlers currently shuffles the order (if a handler at 3rd place get disabled and reenabled, it gets put to the front). Revise the implementation so that the order is preserved (probably by using a list of disable flags rather than moving handlers between two lists).

P.S. I don’t usually like to shout out that I agree on something, but I think the buffer/mode slots answered partly a problem that puzzled me for a long time: http://lambda-the-ultimate.org/node/5630

It turns out that the ritual Emacs modes performs is probably just simulating the Nyxt model!

**Comment by Ambrevar** [comment]
At first glance I agree with proposal 1, namely that we should accept symbols as handlers.

Allow me to sum up the design decision of Nyxt: we aim for a _functional behaviour_ (in the _functional programming_ sense), so in that sense we are very much unllke Emacs.  The rationale is simple: if you can be local you can also be global, but the opposite is not true.

If your issue is with updating existing objects (which requires looping), it's easy to fix with some helper functions, so all we need to do is provide these functions to the user.  This has been annoying everyone for a long time, so let's fix it! :)

Conversely, changing the architecture to be less functional would render locality impossible.

Your last point 5 seem valid as well, but we need to think through what exactly "mode order" means.  Indeed, when a mode is (re)enabled, it runs some code and the result depends on the current buffer, thus on the currently enabled mode and thus on the load order.  This is never really functional, so it kind of makes sense that the re-enabled mode is moved to the front as it indicates when it was enabled.
I guess your issue is with binding priorities.  This is an orthogonal concern, but just as valid.  Maybe we need 2 "mode orders"?

All in all, I can see at least 3 points we may all agree need tackling:
1\. Symbol handlers
5\. Mode order
6\. Existing buffer updater helpers

Does anyone want to send patches about them now?  Otherwise we can open separate issues since this one has dragged one for a while already :)

**Comment by BlueFlo0d** [comment]
> Allow me to sum up the design decision of Nyxt: we aim for a _functional behaviour_ (in the _functional programming_ sense), so in that sense we are very much unllke Emacs. The rationale is simple: if you can be local you can also be global, but the opposite is not true.

Adding a symbol indirection does not affect locality of programmatic part, as long as the programs are written in a sane way. See my example of Smalltalk's object id table. Also, hardly anyone complain about Lisp's `eq` identity, and the default behavior for every function call form. They're both *technically* global! Adding symbol indirection to hooks (and, some other `#'` I've spotted in Nyxt code base) is no more global than using `eq`, or calling a function as `(function . arguments)`, or referring to a class by symbol so that `defclass` updates it, in Common Lisp.

On the other hand, I must voice my concern on your seemly innocent slogan. There's a lots of ambiguity in interpreting it, but at least an interpretation that rejects symbol indirection, is problematic. You can never simulate full flexibility and discoverability without global access across *all meta-level*! On the other hand, the very semantics of Common Lisp itself contradicts this interpretation in many aspects, as the examples I've given above. What language this interpretation leads to? JavaScript. The standard practice in JavaScript is to build programs as nesting of opaque closures. This hinders flexibility and discoverability so much that Emacs still survive in hacker community even if web technology runs faster, has much more libraries, more portable, has more functionality ...! If Nyxt wants to go (traditional) hacker oriented, it must not go down this road.

> If your issue is with updating existing objects (which requires looping), it's easy to fix with some helper functions, so all we need to do is provide these functions to the user. This has been annoying everyone for a long time, so let's fix it! :)

The best way to implement this is to add one level of indirection. It would be very tedious to truly simulate the behavior of symbol indirection (i.e. when a function is redefined, all use of it are updated consistently and implicitly). And to simulate it, you must need to maintain global state to keep track of dependency (buffer list, and what function has hooks defined) as well. A even worse solution will be to require user to keep track of such inevitable global state mentally.

> Conversely, changing the architecture to be less functional would render locality impossible.

As my first paragraph and previous replies suggests, adding a symbol indirection doesn't make the architecture any less functional, it makes it more flexible.

> Does anyone want to send patches about them now? Otherwise we can open separate issues since this one has dragged one for a while already :)

I will likely do them in one batch in a while.

**Comment by BlueFlo0d** [comment]
> Your last point 5 seem valid as well, but we need to think through what exactly "mode order" means. Indeed, when a mode is (re)enabled, it runs some code and the result depends on the current buffer, thus on the currently enabled mode and thus on the load order. This is never really functional, so it kind of makes sense that the re-enabled mode is moved to the front as it indicates when it was enabled. I guess your issue is with binding priorities. This is an orthogonal concern, but just as valid. Maybe we need 2 "mode orders"?

Seems that currently the handler disable/re-enable facility is not used in mode toggling (they just add or remove handlers). So handler order is not really related to mode order.

I don't really see any internal use of this facility in Nyxt yet, but I do think it's the right step and could be used to improve error handling (emacs usually just remove the handler, but disabling it is better). In such case I think it make much more sense if disable then re-enable is a no-op. It will be very useful if a bug happens in a handler (which automatically disables it), the user fix the bug (by redefining the function, so this work only when handler is a symbol!), and re-enable the handler. The state of the handler should returns to the one before the bug trigger.

**Comment by Ambrevar** [comment]
I think you misunderstood me: the functional aspect of the hooks has
nothing to do with the handlers being functions or symbols.

I'm completely fine with having symbols as handlers.

My comment about mode order is also unrelated to handler order.

**Comment by BlueFlo0d** [comment]
> I think you misunderstood me: the functional aspect of the hooks has nothing to do with the handlers being functions or symbols. 

Sorry if that's the case. I've tried to reread your message. May I request some clarification? "functional programming" has been referred to different things during different historical periods, so I think maybe that's the source of confusion:

1. It could mean "higher-order procedures". This is the reason why Lisp was called "functional". This is probably not relevant in the discussion as we all agree on it.
2. It could mean "pure function", or functions whose effect is only dependent on the *extensional* identities of its arguments, as in Haskell. This is why Lisp is no longer called "functional".
3. It could mean state access are somewhat localized. Or, in capability speaking, avoidance of powerful/ambient authority. Or, in layman's term, functions should access only a limited amount of shared mutable states. I feel like this is the interpretation that fits the context the most, although it's not usually referred as "functional".

Under interpretation 3, in many aspects Nyxt is not so much different from Emacs. Encouraging named functions, and even the advice facility, doesn't contradict with interpretation 3 as long as "programmatic components" does not manipulate such facilities directly. AFAIU the difference between Nyxt and Emacs might be that Nyxt want to avoid lots of unnecessary access to shared mutable state in Emacs, like the temporarily moving around the point or mutating the buffer. Is that the case?

> My comment about mode order is also unrelated to handler order.
I get it now. I think the current mode order behavior makes sense (my proposal was about handler order), because of exactly the reason you mentioned. Now you say that it does feel a bit ad-hoc, it's like letting `defmethod` order determined solely by the time of definition. The principled approach would probably be sorting the modes according to a defined partial order, akin to CLOS methods?

**Comment by Ambrevar** [comment]
> AFAIU the difference between Nyxt and Emacs might be that Nyxt want to
> avoid lots of unnecessary access to shared mutable state in Emacs,
> like the temporarily moving around the point or mutating the
> buffer. Is that the case?

Yes :)

> The principled approach would probably be sorting the modes according
> to a defined partial order, akin to CLOS methods?

Sorting could be an option, but I wonder how usable that would be on the
user end.  We must keep in mind that users may use third-party
extensions which are not necessarily aware of each other.

**Comment by BlueFlo0d** [comment]
Before closing this, just want to tell you about a new package I just wrote `https://github.com/BlueFlo0d/named-closure`. I think it is the right way to use closure as handlers. It solves the pain points for closure in general, rather than for the specific use case of handlers, but it also just happen to cover handler use case well. Maybe Nyxt will benefit from replacing handler+anonymous functions with named closures. What do you think?

**Comment by Ambrevar** [comment]
Nice! I've shared some comments.  I believe we can tune it and use it
everywhere in nhooks and Nyxt in the future!

**Comment by Ambrevar** [comment]
@BlueFlo0d OK to close then?
